File: examples/rust/examples/batch_operations.rs
use serde_json::{json, Value};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let api_key =
        env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".to_string());

    // Step 1: Get authentication token
    let client = reqwest::Client::new();
    let token_response = client
        .post(&format!("{}/api/auth/token", base_url))
        .json(&json!({ "api_key": api_key }))
        .send()
        .await?;

    let token_data: Value = token_response.json().await?;
    let token = token_data["token"].as_str().unwrap();
    println!("✓ Authentication successful");

    // Example 1: Batch Insert
    println!("\n=== Batch Insert ===");
    let records = vec![
        json!({"name": "User 1", "email": "user1@example.com", "age": 25}),
        json!({"name": "User 2", "email": "user2@example.com", "age": 30}),
        json!({"name": "User 3", "email": "user3@example.com", "age": 35}),
        json!({"name": "User 4", "email": "user4@example.com", "age": 40}),
        json!({"name": "User 5", "email": "user5@example.com", "age": 45}),
    ];

    let inserts: Vec<Value> = records.iter().map(|r| json!({"data": r})).collect();
    let batch_insert_data = json!({
        "inserts": inserts,
        "bypass_ripple": false
    });

    let insert_response = client
        .post(&format!("{}/api/batch/insert/batch_users", base_url))
        .header("Authorization", format!("Bearer {}", token))
        .json(&batch_insert_data)
        .send()
        .await?;

    let insert_result: Value = insert_response.json().await?;
    let inserted_count = insert_result["successful"]
        .as_array()
        .map(|a| a.len())
        .unwrap_or(0);
    println!("✓ Batch inserted {} records", inserted_count);

    // Verify the inserts
    let verify_response = client
        .post(&format!("{}/api/find/batch_users", base_url))
        .header("Authorization", format!("Bearer {}", token))
        .json(&json!({"limit": 100}))
        .send()
        .await?;

    let all_docs: Value = verify_response.json().await?;
    let total_count = all_docs.as_array().map(|a| a.len()).unwrap_or(0);
    println!(
        "✓ Verified: Found {} total records in collection",
        total_count
    );

    // Example 2: Create test records for update/delete
    println!("\n=== Creating test records for update/delete ===");
    let mut ids = Vec::new();
    for i in 0..3 {
        let response = client
            .post(&format!("{}/api/insert/batch_users", base_url))
            .header("Authorization", format!("Bearer {}", token))
            .json(&json!({"name": format!("Test User {}", i), "value": i}))
            .send()
            .await?;

        let doc: Value = response.json().await?;
        ids.push(doc["id"].as_str().unwrap().to_string());
    }
    println!("Created {} test records", ids.len());

    // Example 3: Batch Update
    println!("\n=== Batch Update ===");
    let updates: Vec<Value> = ids
        .iter()
        .enumerate()
        .map(|(i, id)| {
            json!({
                "id": id,
                "data": {"name": format!("Updated User {}", i), "status": "active"}
            })
        })
        .collect();

    let batch_update_data = json!({
        "updates": updates,
        "bypass_ripple": false
    });

    let update_response = client
        .put(&format!("{}/api/batch/update/batch_users", base_url))
        .header("Authorization", format!("Bearer {}", token))
        .json(&batch_update_data)
        .send()
        .await?;

    let update_result: Value = update_response.json().await?;
    let updated_count = update_result["successful"]
        .as_array()
        .map(|a| a.len())
        .unwrap_or(0);
    println!("✓ Batch updated {} records", updated_count);

    // Verify the updates
    let verify_update = client
        .get(&format!("{}/api/find/batch_users/{}", base_url, ids[0]))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await?;

    let updated_doc: Value = verify_update.json().await?;
    let status = updated_doc["status"]
        .as_str()
        .or_else(|| updated_doc["status"]["value"].as_str())
        .unwrap_or("active");
    println!("✓ Verified: Record updated with status=\"{}\"", status);

    // Example 4: Batch Delete
    println!("\n=== Batch Delete ===");
    let deletes: Vec<Value> = ids.iter().map(|id| json!({"id": id})).collect();
    let batch_delete_data = json!({
        "deletes": deletes,
        "bypass_ripple": false
    });

    let delete_response = client
        .delete(&format!("{}/api/batch/delete/batch_users", base_url))
        .header("Authorization", format!("Bearer {}", token))
        .json(&batch_delete_data)
        .send()
        .await?;

    let delete_result: Value = delete_response.json().await?;
    let deleted_count = delete_result["successful"]
        .as_array()
        .map(|a| a.len())
        .unwrap_or(0);
    println!("✓ Batch deleted {} records", deleted_count);

    // Verify the deletes
    let verify_delete = client
        .get(&format!("{}/api/find/batch_users/{}", base_url, ids[0]))
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await?;

    if verify_delete.status() == 404 {
        println!("✓ Verified: Records successfully deleted (not found)");
    } else {
        println!("✗ Warning: Record still exists after delete!");
    }

    println!("\n✓ All batch operations completed successfully");

    Ok(())
}

File: examples/go/batch_operations.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"

	"github.com/joho/godotenv"
)

var (
	baseURL   string
	apiKey    string
	authToken string
)

func init() {
	godotenv.Load()
	baseURL = os.Getenv("API_BASE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:8080"
	}
	apiKey = os.Getenv("API_BASE_KEY")
	if apiKey == "" {
		apiKey = "a-test-api-key-from-ekodb"
	}
}

func getAuthToken() (string, error) {
	if authToken != "" {
		return authToken, nil
	}

	body, _ := json.Marshal(map[string]string{"api_key": apiKey})
	resp, err := http.Post(baseURL+"/api/auth/token", "application/json", bytes.NewBuffer(body))
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result map[string]interface{}
	json.NewDecoder(resp.Body).Decode(&result)
	authToken = result["token"].(string)
	return authToken, nil
}

func request(method, path string, body interface{}) (map[string]interface{}, error) {
	token, err := getAuthToken()
	if err != nil {
		return nil, err
	}

	var reqBody io.Reader
	if body != nil {
		jsonBody, _ := json.Marshal(body)
		reqBody = bytes.NewBuffer(jsonBody)
	}

	req, _ := http.NewRequest(method, baseURL+path, reqBody)
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == 404 {
		return nil, nil
	}

	respBody, _ := io.ReadAll(resp.Body)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(respBody))
	}
	if len(respBody) == 0 {
		return map[string]interface{}{}, nil
	}

	var result map[string]interface{}
	if err := json.Unmarshal(respBody, &result); err != nil {
		return nil, fmt.Errorf("decode response: %w", err)
	}
	return result, nil
}

func run() (runErr error) {
	fmt.Print("=== Batch Operations (Direct HTTP) ===\n\n")

	_, err := getAuthToken()
	if err != nil {
		return fmt.Errorf("auth failed: %w", err)
	}
	fmt.Println("✓ Authentication successful")
	if _, err := request("DELETE", "/api/collections/batch_users_go", nil); err != nil {
		return fmt.Errorf("initial cleanup failed: %w", err)
	}
	defer func() {
		if _, err := request("DELETE", "/api/collections/batch_users_go", nil); err != nil {
			runErr = errors.Join(runErr, fmt.Errorf("cleanup failed: %w", err))
		}
	}()

	// Example 1: Batch Insert
	fmt.Println("\n=== Batch Insert ===")
	records := []map[string]interface{}{
		{"name": "User 1", "email": "user1@example.com", "age": 25},
		{"name": "User 2", "email": "user2@example.com", "age": 30},
		{"name": "User 3", "email": "user3@example.com", "age": 35},
		{"name": "User 4", "email": "user4@example.com", "age": 40},
		{"name": "User 5", "email": "user5@example.com", "age": 45},
	}

	inserts := make([]map[string]interface{}, len(records))
	for i, r := range records {
		inserts[i] = map[string]interface{}{"data": r}
	}

	batchInsertData := map[string]interface{}{
		"inserts":       inserts,
		"bypass_ripple": false,
	}

	insertResult, err := request("POST", "/api/batch/insert/batch_users_go", batchInsertData)
	if err != nil {
		return fmt.Errorf("batch insert: %w", err)
	}
	insertedCount := 0
	if successful, ok := insertResult["successful"].([]interface{}); ok {
		insertedCount = len(successful)
	}
	fmt.Printf("✓ Batch inserted %d records\n", insertedCount)
	if insertedCount != len(records) {
		return fmt.Errorf("expected %d successful inserts, got %d", len(records), insertedCount)
	}

	// Example 2: Create test records for update/delete
	fmt.Println("\n=== Creating test records for update/delete ===")
	var ids []string
	for i := 0; i < 3; i++ {
		doc, err := request("POST", "/api/insert/batch_users_go", map[string]interface{}{
			"name":  fmt.Sprintf("Test User %d", i),
			"value": i,
		})
		if err != nil {
			return fmt.Errorf("create test record %d: %w", i, err)
		}
		id, ok := doc["id"].(string)
		if !ok || id == "" {
			return fmt.Errorf("create test record %d returned no id", i)
		}
		ids = append(ids, id)
	}
	fmt.Printf("Created %d test records\n", len(ids))

	// Example 3: Batch Update
	fmt.Println("\n=== Batch Update ===")
	updates := make([]map[string]interface{}, len(ids))
	for i, id := range ids {
		updates[i] = map[string]interface{}{
			"id":   id,
			"data": map[string]interface{}{"name": fmt.Sprintf("Updated User %d", i), "status": "active"},
		}
	}

	batchUpdateData := map[string]interface{}{
		"updates":       updates,
		"bypass_ripple": false,
	}

	updateResult, err := request("PUT", "/api/batch/update/batch_users_go", batchUpdateData)
	if err != nil {
		return fmt.Errorf("batch update: %w", err)
	}
	updatedCount := 0
	if successful, ok := updateResult["successful"].([]interface{}); ok {
		updatedCount = len(successful)
	}
	fmt.Printf("✓ Batch updated %d records\n", updatedCount)
	if updatedCount != len(ids) {
		return fmt.Errorf("expected %d successful updates, got %d", len(ids), updatedCount)
	}

	// Example 4: Batch Delete
	fmt.Println("\n=== Batch Delete ===")
	deletes := make([]map[string]interface{}, len(ids))
	for i, id := range ids {
		deletes[i] = map[string]interface{}{"id": id}
	}

	batchDeleteData := map[string]interface{}{
		"deletes":       deletes,
		"bypass_ripple": false,
	}

	deleteResult, err := request("DELETE", "/api/batch/delete/batch_users_go", batchDeleteData)
	if err != nil {
		return fmt.Errorf("batch delete: %w", err)
	}
	deletedCount := 0
	if successful, ok := deleteResult["successful"].([]interface{}); ok {
		deletedCount = len(successful)
	}
	fmt.Printf("✓ Batch deleted %d records\n", deletedCount)
	if deletedCount != len(ids) {
		return fmt.Errorf("expected %d successful deletes, got %d", len(ids), deletedCount)
	}

	// Verify the deletes
	verifyDelete, err := request("GET", "/api/find/batch_users_go/"+ids[0], nil)
	if err != nil {
		return fmt.Errorf("verify delete: %w", err)
	}
	if verifyDelete == nil {
		fmt.Println("✓ Verified: Records successfully deleted (not found)")
	} else {
		return fmt.Errorf("record still exists after delete")
	}

	return nil
}

func main() {
	if err := run(); err != nil {
		fmt.Printf("Error: %v\n", err)
		os.Exit(1)
	}
	fmt.Println("\n✓ All batch operations completed successfully")
}
